3-enrich/stream-enrich/src/test/scala/com.snowplowanalytics.snowplow.enrich.stream/KafkaIntegrationSpec.scala

```diff
@@ -0,0 +1,128 @@
+    configuration: EnrichConfig,
+    resolver: Resolver,
+    registry: EnrichmentRegistry,
+    tracker: Option[Tracker]): Future[Unit] = Future {
+    EnrichApp.run(configuration, resolver, registry, tracker)
+  }
+
+  def producerTimeoutSec: Int
+  def inputProduced = Try { Await.result(produce, Duration(s"$producerTimeoutSec sec")) }
+
+  def produce: Future[Unit] = Future {
+    val testKafkaPropertiesProducer = {
+      val props = new Properties()
+      props.put("bootstrap.servers", kafkaHost)
+      props.put("client.id", "producer-george")
+      props.put("key.serializer", "org.apache.kafka.common.serialization.StringSerializer")
+      props.put("value.serializer", "org.apache.kafka.common.serialization.ByteArraySerializer")
+      props
+    }
+    val testProducer = new KafkaProducer[String, Array[Byte]](testKafkaPropertiesProducer)
+
+    val events = inputGood
+    events.foreach { r =>
+      testProducer.send(new ProducerRecord(testGoodIn, "key", r))
+    }
+    testProducer.flush
+    testProducer.close
+  }
+  private def getListOfRecords(cr: ConsumerRecords[String, String]): List[String] =
+    cr.map(_.value).toList
+
+  def getRecords(topic: String, expectedRecords: Int, timeoutSec: Int): Future[List[String]] =
+    Future {
+      val started = System.currentTimeMillis
+      val testKafkaPropertiesConsumer = {
+        val props = new Properties()
+        props.put("bootstrap.servers", kafkaHost)
+        props.put("auto.offset.reset", "earliest")
+        props.put("key.deserializer", "org.apache.kafka.common.serialization.StringDeserializer")
+        props
+          .put("value.deserializer", "org.apache.kafka.common.serialization.StringDeserializer")
+        props.put("group.id", s"consumer-$topic")
+        props
+      }
+      val testConsumerPii = new KafkaConsumer[String, String](testKafkaPropertiesConsumer)
+      testConsumerPii.subscribe(List(topic))
+      var records = getListOfRecords(testConsumerPii.poll(100))
+      while (((System.currentTimeMillis - started) / 1000 < timeoutSec - 1) && records.size < expectedRecords) {
+        records = records ++ getListOfRecords(testConsumerPii.poll(100))
+      }
+      records
+    }
+
+  def consumerExecutionTimeoutSec: Int
+  def producedBadRecords: Future[List[String]] =
+    getRecords(testBad, expectedBad, consumerExecutionTimeoutSec)
+  def producedGoodRecords: Future[List[String]] =
+    getRecords(testGood, expectedGood, consumerExecutionTimeoutSec)
+  def producedPiiRecords: Future[List[String]] =
+    getRecords(testPii, expectedPii, consumerExecutionTimeoutSec)
+  def allFutures: Future[(List[String], List[String], List[String])] =
+    for {
+      good <- producedGoodRecords
+      bad  <- producedBadRecords
+      pii  <- producedPiiRecords
+    } yield (good, bad, pii)
+  def outputResult: Try[(List[String], List[String], List[String])] = Try {
+    Await.result(allFutures, Duration(s"$consumerExecutionTimeoutSec sec"))
+  }
```

Comment:
can't we do away with the `Try` / `Await.result` as it defeats the purpose of having futures in the first place?

we can maybe let specs2 do the matching directly on futures?

Reply:
Sure. So that I understand the issue, can you please explain how it defeats the purpose of having futures in the first place?

Reply:
You're building a future to just await its results synchronously so you lose the purpose of the future (async computations) and you wrap the result in a try, losing Failure / Success resulting of the future.

You might as well do without the future if you're doing synchronous computations wrapped in a try / catch.

Do you see what I mean?

Reply:
I don't think I do understand.
Currently there are three parallel background tasks that will complete when they receive the number of results that they are expecting. As this may never happen there needs to be a timeout. I am sure you realise that the future returns before the timeout when the result is ready. If we were waiting for the three results synchronously we would need to wait for at least as long as the timeout (45 sec). Actually the consumers return in about 2 sec. 
So the objective here is to wait for a result that we don't know when it is coming (but usually pretty quickly in the successful case), but not wait forever, to do it in parallel for the three results expected so as not to delay the test unnecessarily. If the code in the future throws an exception, I see that exception in the Try. If the code times out it throws so the try is needed there. 
These results are needed to complete the test evaluation so they need to be awaited. 
I am happy to use the specs2 matcher (which looks almost identical in code):

```
  /**
   * when a Future contains a result, it can be awaited to return this result
   */
  implicit class futureAsResult[T : AsResult](f: Future[T]) {
    def await: Result = await()
    def await(retries: Int = 0, timeout: FiniteDuration = 1.seconds): Result = {
      def awaitFor(retries: Int, totalDuration: FiniteDuration = 0.seconds): Result = {
        try Await.result(f.map(value => AsResult(value)), timeout)
        catch {
          case e: TimeoutException =>
            if (retries <= 0) Failure(s"Timeout after ${totalDuration + timeout}")
            else awaitFor(retries - 1, totalDuration + timeout)
          case other: Throwable    => throw other
        }
      }
      awaitFor(retries)
    }
  }
```

however since I still don't understand what you mean by "lose the purpose of the future" I don't know if it will address your concern. I think this will be easier over HipChat..

Reply:
Recapping @BenFradet we will use specs2 Matcher.await instead of Await.result to give the option to not block to whoever is implementing another kafka test in the future.

Reply:
yes since specs2 doesn't do async testing apparently...